src/components/world/EventHelper.tsx

```diff
@@ -0,0 +1,118 @@
+import React from "react";
+import {
+  EVENT_CAMERA_MOVE_TO,
+  EVENT_ACTOR_MOVE_TO,
+  EVENT_ACTOR_SET_POSITION,
+  EVENT_OVERLAY_SHOW,
+  EVENT_OVERLAY_MOVE_TO,
+  EVENT_IF_ACTOR_AT_POSITION,
+} from "lib/compiler/eventTypes";
+
+const TILE_SIZE = 8;
+
+type Arg<T> = {
+  value: T,
+  type: "variable" | "property" | string
+};
+
+// This is grossly over-typed
+const argValue = <T extends unknown>(arg: T):
+  T extends Arg<infer R> ?
+    T['type'] extends 'variable' | 'property' ? undefined : R:
+    T => {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const isShapedArg = (v: any): v is Arg<any> => v?.value;
+  if (isShapedArg(arg)) {
+    if (arg.type === "variable" || arg.type === "property") {
+      return undefined as never;
+    }
+    return arg.value;
+  }
+  return arg as never;
+};
```

Comment:
This code I wrote is frankly disgusting. I sort [nerd sniped](https://xkcd.com/356/) myself on this one. However, it's also _technically_ correct ([the best kind of correct](https://www.youtube.com/watch?v=hou0lU8WMgo)) and is the most type-strict variation of this logic present.

In particular, this line:

```
T['type'] extends 'variable' | 'property' ? undefined : R
```

Will literally not run unless an `arg` is defined with `as const` (or if the root typing is strict and not just assigned to `string`).

For this reason, no problem if we want to cull the typings down to something more reasonably maintainable. Apologies for even committing this grossness